File: examples/ruby_on_rails_ruby_graphql/app/graphql/mutations/create_user.rb
# frozen_string_literal: true

module Mutations
  class CreateUser < BaseMutation
    field :user, Types::UserType, null: false

    argument :email, String, required: true
    argument :name, String, required: true
    argument :password, String, required: true

    def resolve(email:, name:, password:)
      user = User.create! email: email,
                          name: name,
                          password: password
      {
        user: user
      }
    end
  end
end

File: examples/go/graph/resolvers/mutation/createUser.go
package mutation

import (
	"fmt"

	"github.com/adeynack/graphql_examples/examples/go/graph/model"
	"github.com/adeynack/graphql_examples/examples/go/graph/resolvers/gqlpolicy"
	"github.com/adeynack/graphql_examples/examples/go/service"
)

func CreateUser(ctx service.ReqCtx, input model.CreateUserInput) (*model.CreateUserResponse, error) {
	response := &model.CreateUserResponse{ClientMutationID: input.ClientMutationID}
	if err := gqlpolicy.Authenticated(ctx); err != nil {
		return response, err
	}
	response.User = &model.User{
		Email: input.Email,
		Name:  input.Name,
	}
	if err := response.User.SetPassword(ctx.ServerSalt, input.Password); err != nil {
		return response, fmt.Errorf("error setting new user's password: %v", err)
	}
	if err := ctx.DB.Create(response.User).Error; err != nil {
		return response, fmt.Errorf("error creating new user: %v", err)
	}
	return response, nil
}
